### PR TITLE
WIP: Breakpoint management for runspaces

### PR DESCRIPTION
**Pull request by KirkMunro**
# PR Summary

- Adds an experimental flag called "Microsoft.PowerShell.Utility.PSManageBreakpointsInRunspace". All experimental functionality described below is enabled/disabled using that flag.
- Adds an experimental `-Runspace` parameter `Get-PSBreakpoint`, `Set-PSBreakpoint`, `Enable-PSBreakpoint`, `Disable-PSBreakpoint`, and `Remove-PSBreakpoint`.
- Refactors the classes behind the `*-PSBreakpoint` cmdlets so that they all work consistently. This refactoring did not change the behavior of these cmdlets beyond adding support for runspace breakpoint management.
- Removes the recently added `New-PSBreakpoint` experimental cmdlet in favor of working with the older, more well known `*-PSBreakpoint` cmdlets for runspace breakpoint management.
- Refactors `DebugRunspaceCommand` and `DebugJobCommand` so that they both derive from a common `PSRemoteDebugCmdlet` base class. As part of this refactoring, some of the resource strings have been moved to be co-located with the new base class where they are used.
- Fixes a very minor issue where a remote breakpoint would not be displayed properly in the host when entering a remote debugger.
- Adds `SetCommandBreakpoint`, `SetVariableBreakpoint`, `SetLineBreakpoint`, `EnableBreakpoint`, `DisableBreakpoint`, and `RemoveBreakpoint` public APIs to the `Debugger` class, plumbs them through the engine, and hooks them up for remote debugger work. Also hooks up the `GetBreakpoint`, `GetBreakpoints`, and `SetBreakpoints` `Debugger` class APIs for remote debugger work.
- Adds a `-NoInitialBreak` experimental parameter to `Debug-Runspace` and `Debug-Job` that allows users to attach the debugger without invoking a `BreakAll`. The corresponding methods for the same have boolean parameters allowing this as well.

## PR Context

This PR was opened in response to [this discussion](https://github.com/PowerShell/PowerShell/pull/8923#issuecomment-503267596) between @TylerLeonhardt, @PaulHigin and I.

It helps because:
- It allows scripters to fully manage breakpoints in a remote runspace without having to use `Debug-*` to attach the debugger to that runspace. This provides a better, more PowerShell-oriented approach to debugging.
- It allows @TylerLeonhardt to attach a debugger to a runspace without invoking a `BreakAll` in PSES.
- It simplifies the implementation behind `*-PSBreakpoint` cmdlets.
- The implementation behind `Debug-Runspace` and `Debug-Job`, as defined in `DebugRunspaceCommand` and `DebugJobCommand`, respectively, had a lot of duplicated logic and a few unintentional differences. Moving these classes into a shared base class structure eliminates the unintentional differences and removes the duplicated code.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] Some (experimental features are required if breaking changes are included).
- **Experimental features**
    - [X] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [X] Experimental feature name(s): Microsoft.PowerShell.Utility.PSManageBreakpointsInRunspace
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [X] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/resources/DebuggerStrings.resx [hunk 1]
```diff
@@ -246,4 +246,25 @@ The current session does not support debugging; operation will continue.
...
+  <data name="ScriptCompletedMessage" xml:space="preserve">
+    <value>Command or script completed.</value>
+  </data>
+  <data name="DebuggingNoDebugger" xml:space="preserve">
+    <value>Cannot debug because the host has no debugger. Try debugging inside the PowerShell console or with Visual Studio Code, both of which have built-in debuggers.</value>
+  </data>
+  <data name="DebuggingNoHost" xml:space="preserve">
+    <value>Cannot debug because there is no host or host UI. The debugger requires a host and host UI for debugging.</value>
+  </data>
+</root>
```
Comment (iSazonov) [inline src/System.Management.Automation/resources/DebuggerStrings.resx]: Please fix all resx files in the PR - I see EOL here and extra spaces.

**Comment by iSazonov** [comment]
@KirkMunro Is it possible to split the PR on some small ones? The PR is so large that we can wait too long for approval.

File: src/System.Management.Automation/resources/DebuggerStrings.resx (hunk 1 above)
Comment (KirkMunro) [reply src/System.Management.Automation/resources/DebuggerStrings.resx]: I didn't make any of those changes. I don't edit these by hand (with one exception -- there was a merge conflict, and I resolved that by hand, but it didn't impact the newline nor the spaces). Further, the other resx file does not have an extra EOL at the end of the file either, but it's not called out because the last entry didn't change. I think these come from tooling (I use Visual Studio 2019 latest update for resource string changes), so I'm not sure what should be done here.

**Comment by KirkMunro** [comment]
> @KirkMunro Is it possible to split the PR on some small ones? The PR is so large that we can wait too long for approval.

That sounds like quite a bit more work, with little tangible value; however, a "review map" might help. Here is a break down of the changes in each file, with like changes grouped together to allow them to be reviewed more easily. Between these extra details and being able to mark files as viewed in the GitHub UI, I think it should be much easier for others to review this PR.

### Cleanup of changes done in @TylerLeonhardt's other PR that this supercedes

`src/Microsoft.PowerShell.Commands.Utility/commands/utility/New-PSBreakpoint.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/EnableDisableRunspaceDebugCommand.cs`
- simple removal of changes added by @TylerLeonhardt in PR #; instead of
creating breakpoints using `New-PSBreakpoint` and passing breakpoints into
`Enable-RunspaceDebug`, breakpoints are managed on a runspace using the
original set of `*-PSBreakpoint` cmdlets.

`src/Microsoft.PowerShell.Commands.Utility/commands/utility/PSBreakpointCreationBase.cs`
- simple removal of previous refactoring work that was done in the same PR

`src/Modules/Unix/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1`
`src/Modules/Windows/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1`
- removed `New-PSBreakpoint` reference
- changed experimental feature name/description to match what it actually does now
(this was previously missing from the Windows version of the file)

`src/System.Management.Automation/engine/InitialSessionState.cs`
- removed `New-PSBreakpoint` reference

`test/powershell/engine/Basic/DefaultCommands.Tests.ps1`
- removed `New-PSBreakpoint` reference

`test/powershell/Modules/Microsoft.PowerShell.Utility/Enable-RunspaceDebug.Tests.ps1`
- removed this file as it was based on the other PR that this supercedes

### Changes for Pester tests

`test/powershell/Modules/Microsoft.PowerShell.Utility/RunspaceBreakpointManagement.Tests.ps1`
- added tests for runspace breakpoint management (both local and remote runspaces)

`test/tools/TestMetadata.json`
- updated to match new experimental feature name and test file

### Changes to support `-Runspace` parameter in `*-PSBreakpoint` commands

`src/Microsoft.PowerShell.Commands.Utility/commands/utility/RunspaceParameter.cs`
- defined a transformation attribute that converts integers, GUIDs and strings into
runspaces; this is for use in the `-Runspace` parameter added to the `*-PSBreakpoint`
cmdlets updated below (it keeps the number of parameter sets low in those commands
while still supporting multiple ways to reference a runspace)

`src/Microsoft.PowerShell.Commands.Utility/commands/utility/PSBreakpointCommandBase.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/PSBreakpointAccessorCommandBase.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/Get-PSBreakpoint.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/Set-PSBreakpoint.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/PSBreakpointUpdaterCommandBase.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/Disable-PSBreakpoint.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/Enable-PSBreakpoint.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/Remove-PSBreakpoint.cs`
- added `-Runspace` parameter to each `*-PSBreakpoint` cmdlet
- standardized parameter set names across this set of cmdlets
- refactored `*-PSBreakpoint` cmdlets into class hierarchy, as follows:
    ```none
                           PSBreakpointCommandBase
                              /              \
    PSBreakpointAccessorCommandBase        PSBreakpointUpdaterCommandBase
                 |                                       |
     Get/SetPSBreakpointCommand        Disable/Enable/RemovePSBreakpointCommand
    ```

### Changes to support managing breakpoints in debugger via the SDK

`src/System.Management.Automation/engine/debugger/Breakpoint.cs`
- fixed minor issue in display of breakpoint that was hit when attaching
the debugger to a remote runspace
- updated RemoveSelf method to return a bool if a breakpoint was removed

`src/System.Management.Automation/engine/debugger/debugger.cs`
- added SetCommandBreakpoint, SetLineBreakpoint, SetVariableBreakpoint,
RemoveBreakpoint, EnableBreakpoint, and DisableBreakpoint public methods
in Debugger class
- added DebugJob method that allows you to attach a debugger to a job
without invoking a BreakAll
- changed return type of AddCommandBreakpoint, AddLineBreakpoint and
AddVariableBreakpoint on ScriptDebugger class to be specific to the type
of breakpoint they add
- removed New*Breakpoint methods on ScriptDebugger class (these have been
superceded by the Set*Breakpoint public methods)
- updated Remove*Breakpoint methods to return a bool indicating whether or
not a breakpoint was removed
- defined overrides in ScriptDebugger for the *Breakpoint methods that were
added to the base class, and grouped them together (this involved replacing
some methods that were previously internal)
- moved a few methods into proper groupings according to existing regions
- defined overrides in NestedDebugger for the *Breakpoint methods that were
added to the base class
- added internal virtual function names for breakpoint functions that can now
be invoked in a remote runspace

`src/System.Management.Automation/engine/remoting/client/remoterunspace.cs`
- added overrides to remote debugger for the *Breakpoint methods that were
added to the base class
- added internal helper method to handle invoking those methods remotely,
capturing their results, and processing any errors

`src/System.Management.Automation/engine/remoting/client/Job.cs`
- added IsFinished property to simplify code elsewhere that checks if a job
is finished
- added new *Breakpoint* methods to JobDebugger class and grouped them together

`src/System.Management.Automation/engine/server/ServerRunspacePoolDriver.cs`
- added support for the new *Breakpoint virtual functions to the
PreProcessDebuggerCommand method
- added overrides for *Breakpoint methods to ServerRemoteDebugger class
- added override for new DebugJob overload method to ServerRemoteDebugger class

`src/System.Management.Automation/engine/serialization.cs`
- added `RehydrationFlags.NullValueOk` to `Script` and `Action` properties of
breakpoints (this ensures they are set to null instead of empty when rehydrated,
which matches how they are set before they are dehydrated)

`src/System.Management.Automation/engine/hostifaces/PSDataCollection.cs`
- added AddRange method to simplify adding a collection to a collection

### Addition of `-NoInitialBreak` parameter to `Debug-Job` and `Debug-Runspace`, plus refactoring

`src/Microsoft.PowerShell.Commands.Utility/resources/Debugger.resx`
`src/System.Management.Automation/resources/RemotingErrorIdStrings.resx`
`src/System.Management.Automation/resources/DebuggerStrings.resx`
- moved a bunch of resource strings into src/System.Management.Automation/resources/DebuggerStrings.resx
as part of the refactoring of `DebugRunspaceCommand` and `DebugJobCommand` under
a common base class
- added some resource strings

`src/System.Management.Automation/engine/remoting/commands/PSRemoteDebugCmdlet.cs`
`src/Microsoft.PowerShell.Commands.Utility/commands/utility/DebugRunspaceCommand.cs`
`src/System.Management.Automation/engine/remoting/commands/DebugJob.cs`
- added `-NoInitialBreak` parameter to `Debug-Runspace` and `Debug-Job` cmdlets
- moved common code out of the cmdlet classes into new `PSRemoteDebugCmdlet` class
- refactored these two cmdlets to eliminate unintended differences in the UX they
provided into a class hierarchy, as follows:
    ```none
               PSRemoteDebugCmdlet
                 /              \
    DebugJobCommand        DebugRunspaceCommand
    ```

**Comment by PaulHigin** [comment]
I'll try to take a look at this later this week, but may not get to it until the following week.  Since this is such a large change I will mark it for committee review.

File: src/System.Management.Automation/resources/DebuggerStrings.resx (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/resources/DebuggerStrings.resx]: GitHub diff shows that new spaces was added in many places. Perhaps VS 2019 did accidentally this. But in any case we need revert this.

**Comment by KirkMunro** [comment]
@PaulHigin FYI for when you do get to look at this:

I'm experiencing an intermittent issue with my `RunspaceBreakpointManagement.Tests.ps1` Pester test series where setting a breakpoint in a remote runspace (i.e. one associated with a background job) gets stuck. On my system, when I don't have my VS 2019 debugger connected to the PowerShell session process, I can very often reproduce my intermittent issue the _second_ time that I manually invoke that Pester test file in a single PowerShell session. All tests pass in the first run, but the second run hangs at a certain point (the moment the first MS-PSRDP call is made. Digging in to this on my system, I believe it's getting stuck between the moment it invokes the virtual `__Set-PSBreakpoint` function in a nested PowerShell instance, which should set the breakpoint on the remote runspace. The local process seems to be waiting for the pipeline to finish, and the debugger in the remote (job) process seems to be waiting for a client to connect to it, so this seems to have to do with the MS-PSRDP.

In the PR that this supercedes, @TylerLeonhardt switched many of the collections to their concurrent equivalent, and he updated the breakpoint ID increment to be thread safe as well.

I know little about how the MS-PSRDP plumbing works other than what I learned while working on this PR, so I wanted to let you know what I ran into so that you could have a look when you get to this review work because so far I can't figure out the root cause.

File: src/System.Management.Automation/resources/DebuggerStrings.resx (hunk 1 above)
Comment (KirkMunro) [reply src/System.Management.Automation/resources/DebuggerStrings.resx]: No need to apologize. I'll keep an eye on this one with other PRs where I change resource strings because ultimately, if tooling changes to the point where we need to fight against it, we may need to take on such changes in an independent PR. I'll only suggest that if this becomes an ongoing, persistent issue to deal with when making resource file changes.

Comment (KirkMunro) [reply src/System.Management.Automation/resources/DebuggerStrings.resx]: Also, these unintended whitespace and https->http changes have been reverted.

**Comment by PaulHigin** [comment]
cc @SteveL-MSFT
I haven't had a chance yet to look at this, but should by next week.  However, I agree with @iSazonov that it makes sense to break this into smaller scoped changes.  In particular, the changes to the remoting protocol should be separate.  To support these remoting changes, we need to update the protocol version and update capability to ensure down level compatibility.  In addition, the protocol documents must also be updated with the new protocol version additions.  This is an absolute requirement for Windows.

I don't believe we ported all of the remote debugging tests (cc @JamesWTruher), and this makes it difficult to know if any regressions are introduced.  So we need to be careful of any changes we make to the remoting parts.

**Comment by SteveL-MSFT** [comment]
@PaulHigin can you summarize what exactly you'd like the @PowerShell/powershell-committee to review here?

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this.  We appreciate the effort to help improve the debugging experience, however, this PR as it currently exists is too large (size and scope) and risky (due to lack of our own complete test coverage) to accept.  If we are to review PRs like this in the future, we will need:

  - The changes to be split into multiple PRs:
    - refactoring with adequate test coverage (PS Team has existing debugging tests that can be shared, but needs to be ported from an internal (will not be OSS) test framework to Pester/XUnit)
    - local debugging
    - remote debugging (evaluate if this is necessary due to protocol changes)
  - Protocol changes necessitate a change to the protocol documentation which is not a small effort and should be a separate item and also involves cross remoting version validation

**Comment by KirkMunro** [comment]
>     * PS Team has existing debugging tests that can be shared

When/how can those be shared? I've heard about internal debugging tests a number of times, on other PRs and now on this PR, but up to now they've only been mentioned as something that exists.

**Comment by KirkMunro** [comment]
I'm in the process of breaking this up into multiple PRs, starting with PR #10338.

**Comment by KirkMunro** [comment]
Phase 2 of splitting this up: PR #10492, that includes the `*-PSBreakpoint` cmdlet changes.

**Comment by iSazonov** [comment]
@KirkMunro Please resolve merge conflicts.
